source/distributed/fully_distributed_tria.cc

```diff
@@ -0,0 +1,525 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/memory_consumption.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declarations
+namespace GridGenerator
+{
+  template <int dim, int spacedim>
+  void
+  hyper_cube(Triangulation<dim, spacedim> &tria,
+             const double                  left,
+             const double                  right,
+             const bool                    colorize);
+} // namespace GridGenerator
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    namespace internal
+    {
+      /**
+       * Check if `parent` is the parent of `child`.
+       */
+      template <int dim>
+      bool
+      is_parent(typename CellId::binary_type parent,
+                typename CellId::binary_type child)
+      {
+        if (child[0] != parent[0])
+          return false; // has same coarse cell
+        if ((child[1] >> 2) != ((parent[1] >> 2) + 1))
```

Comment:
can you explain what the shift operator does?

I general, the remainder of the function does a lot of bit fiddling that will be incomprehensible to most everyone. It could use a few high-level comments that explain the overall algorithmic flow.